spec/features/user_authentication_feature_spec.rb

```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.feature 'user authentication' do
+  context 'with authentication' do
+    let(:cso_password) { 'password' }
+    let(:cso_user) { User.create(email: 'cso@example.com', password: cso_password, role: :cso) }
+
+    scenario 'logs in with valid credentials' do
+      visit new_user_session_path
+      fill_in 'user_email', with: cso_user.email
+      fill_in 'user_password', with: cso_password
+      click_button 'Sign in'
+      expect(page).to have_content('CSO DASHBOARD')
+    end
+
+    scenario 'logs in with invalid credentials' do
+      visit new_user_session_path
+      fill_in 'user_email', with: cso_user.email
+      fill_in 'user_password', with: 'notarealpassword'
+      click_button 'Sign in'
+      expect(page).to_not have_content('CSO DASHBOARD')
+    end
+
+    context 'unauthenticated' do
+      scenario 'logs in with valid credentials' do
+        visit root_path
+        expect(page).to have_button('Sign in')
+      end
+    end
+  end
+end
```

Comment:
This example could be made a bit more subjectively readable like this:

``` ruby
RSpec.feature 'user authentication' do
  background do
    User.create!(role: :cso, email: 'cso@example.com', password: 'password')
  end

  context 'when unauthenticated' do
    scenario 'visiting the root url' do
      visit root_path
      expect(page).to have_button('Sign in')
    end
  end

  context 'when signing in' do
    context 'with valid credentials' do
      scenario 'the cso logs in' do
        sign_in 'cso@example.com', 'password'
        expect(page).to have_content('CSO DASHBOARD')
      end
    end

    context 'with invalid credentials' do
      scenario 'logs in with invalid credentials' do
        sign_in 'cso@example.com', 'notapassword'
        expect(page).to_not have_content('CSO DASHBOARD')
      end
    end

    def sign_in(email, password)
      visit new_user_session_path
      fill_in 'user_email', with: email
      fill_in 'user_password', with: password
      click_button 'Sign in'
    end 
  end
end
```

Also adding FactoryGirl allows you to use traits to create your models cleanly, e.g:

``` ruby
FactoryGirl.create :user, :cso, email: cso_email, password: cso_password
```

One more thing - doesn't `create` fail because of the lack of a password confirmation?